<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>envMRMEGAfm: environment-adjusted MR-MEGA Fine-mapping • envMRMEGAfm</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="envMRMEGAfm: environment-adjusted MR-MEGA Fine-mapping">
<meta property="og:description" content="envMRMEGAfm">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">envMRMEGAfm</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/envMRMEGAfm.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>envMRMEGAfm: environment-adjusted MR-MEGA
Fine-mapping</h1>
                        <h4 data-toc-skip class="author">Siru Wang</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:sw2077@cam.ac.uk" class="email">sw2077@cam.ac.uk</a>
      
                  
            <h4 data-toc-skip class="date">05 August, 2025</h4>
      
      
      <div class="hidden name"><code>envMRMEGAfm.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The R package <strong>envMRMEGAfm</strong> (env-MR-MEGA fine-mapping)
is designed for simultaneously fine-mapping genetic associations across
multiple cohorts, allowing for multiple causal variants. This method
only requires cohort-level summary statistics, the corresponding linkage
disequilibrium (LD) matrices, and optionally, cohort-level environmental
covariates (e.g. mean or proportion). Sometimes, we are unable to obtain
LD from each cohort. In such cases, we can estimate LD structure using
reference panels, such as 1000 Genomes <span class="citation">(<a href="#ref-10002015global">Consortium et al. 2015</a>)</span>, from a
genetically similar population to the GWAS. This method is powerful,
versatile and computationally fast in pinpointing potential causal
variants while accounting for differing environmental exposures that are
specific to each cohort, in addition to ancestry.</p>
<p>Accounting for environmental exposures that differ between cohorts
and likely impact the variability of traits, the R package
<strong>envMRMEGAfm</strong> introduces MR-MEGAfm and env-MR-MEGAfm,
building upon MR-MEGA <span class="citation">(<a href="#ref-magi2017trans">Mägi et al. 2017</a>)</span> and env-MR-MEGA
<span class="citation">(<a href="#ref-wang2024accounting">Wang et al.
2024</a>)</span> meta-regression frameworks. In addition to the two key
fine-mapping methods, <strong>envMRMEGAfm</strong> package also includes
the method for estimating conditional and joint genetic variant effects,
which was proposed in GCTA <span class="citation">(<a href="#ref-yang2012conditional">Yang et al. 2012</a>)</span> and the
meta-analysis approaches, <a href="https://genomics.ut.ee/en/tools/" class="external-link">MR-MEGA</a> <span class="citation">(<a href="#ref-magi2017trans">Mägi et al.
2017</a>)</span> and <a href="https://SiruRooney.github.io/env.MRmega/" class="external-link">env-MR-MEGA</a> <span class="citation">(<a href="#ref-wang2024accounting">Wang et al.
2024</a>)</span>.</p>
<p>The main R functions in this package are:</p>
<ol style="list-style-type: decimal">
<li><p><code>env_MR_MEGA_fm</code>:A <strong>fine-mapping</strong>
approach applicable across multiple cohorts, with the flexibility to
account for both ancestral and environmental effects (env-MR-MEGAfm) or
only ancestral effects (MR-MEGAfm).</p></li>
<li><p><code>MR_mega_run</code>: The (environment-adjusted) MR-MEGA
meta-regression model.</p></li>
<li><p><code>meta_sel_condest</code>: The approximate conditional
genetic effects method.</p></li>
<li><p><code>meta_sel_joinest</code>: The approximate joint genetic
effects method.</p></li>
</ol>
<p>This vignette guides readers in fine-mapping the genetic associations
for a trait across multiple cohorts using MR-MEGAfm and env-MR-MEGAfm.
For convenience of illustration, a simulated dataset is provided in the
<strong>envMRMEGAfm</strong> package and details of how to run
(env-)MR-MEGAfm method on your own data are explained.</p>
</div>
<div class="section level2">
<h2 id="preparation-for-env-mr-megafm">Preparation for (env-)MR-MEGAfm<a class="anchor" aria-label="anchor" href="#preparation-for-env-mr-megafm"></a>
</h2>
<p>Before running the (env-)MR-MEGAfm method, several input files need
to be prepared. Each input file must also include mandatory columns
about genetic variants.</p>
<div class="section level3">
<h3 id="input-data-and-format-requirements">Input data and format requirements<a class="anchor" aria-label="anchor" href="#input-data-and-format-requirements"></a>
</h3>
<p>The (env-)MR-MEGAfm starts with a model with the most significant SNP
in the single-SNP meta-analysis with P value below the genome-wide
significant level <span class="math inline">\(5\times 10^{-8}\)</span>.
Therefore, it is required that the single-SNP meta-analysis are
conducted by MR-MEGA <span class="citation">(<a href="#ref-magi2017trans">Mägi et al. 2017</a>)</span> and env-MR-MEGA
<span class="citation">(<a href="#ref-wang2024accounting">Wang et al.
2024</a>)</span>. It is noted that the meta-analysis results from both
MR-MEGA and env-MR-MEGA can be obtained from <code>MR_mega_lrw</code>
and <code>MR_mega_env_lrw</code>, which are wrapper functions in the
<strong>env.MRmega</strong> R package.</p>
<p>For an overview of <strong>env.MRmega</strong> R package, see <a href="https://SiruRooney.github.io/env.MRmega/" class="external-link">env.MRmega</a> GitHub
page.</p>
<ol style="list-style-type: decimal">
<li>A list of GWAS files is required for conducting both the
meta-analysis and the following fine-mapping methods. The format of each
file has mandatory column headers, as in the <a href="https://genomics.ut.ee/en/tools/" class="external-link">MR-MEGA</a> software :</li>
</ol>
<ul>
<li>
<code>MARKERNAME</code> - snp name</li>
<li>
<code>N</code> - sample size</li>
<li>
<code>EAF</code> - effect allele frequency</li>
<li>
<code>EA</code> - effect allele</li>
<li>
<code>NEA</code> - non effect allele</li>
<li>
<code>CHROMOSOME</code> - chromosome of marker</li>
<li>
<code>POSITION</code> - position of marker</li>
<li>
<code>BETA</code> - beta</li>
<li>
<code>SE</code> - std.error</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><p>When fitting the (env-)MR-MEGA model and running (env-)MR-MEGAfm
method, we need to prepare the axes of genetic variation (PCs) and the
cohort-level environment exposures. If PCs are not available,
<code>MR_mega_MDS</code> R function implemented in <a href="https://SiruRooney.github.io/env.MRmega/" class="external-link">env.MRmega</a> package
is used for calculating PCs by inputting the file containing all GWAS
files location and the number of PCs.</p></li>
<li><p>The meta-analysis results, as the second required file, can be
obtained from either <a href="https://genomics.ut.ee/en/tools/" class="external-link">MR-MEGA</a> software or <a href="https://SiruRooney.github.io/env.MRmega/" class="external-link">env.MRmega</a> package.
The output from both approaches must contain the following
columns.</p></li>
</ol>
<ul>
<li><p><code>MARKERNAME</code>: unique marker identification across all
cohorts.</p></li>
<li><p><code>CHROMOSOME</code>: chromosome of marker.</p></li>
<li><p><code>POSITION</code>: physical position in chromosome of
marker.</p></li>
<li><p><code>N</code>: sample size.</p></li>
<li><p><code>chisq_association</code>, <code>ndf_association</code> and
<code>pvalue_association</code>: test of the null hypothesis of no
association of each variant. <code>chisq_</code>, <code>ndf_</code> and
<code>pvalue_</code> refer to chisq value, number of degrees of freedom
and p-values of the corresponding test.</p></li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>In addition to the list of GWAS files and the corresponding
meta-analysis results, the (env-)MR-MEGAfm method requires the
cohort-level LD correlation matrices as a third input file. When the
cohort-level LD correlation is unavailable, an approximation of the LD
correlation matrix based on a reference panel may be used. For each LD
correlation structure, both the column names and row names correspond to
<code>MARKERNAME</code>.</li>
</ol>
</div>
<div class="section level3">
<h3 id="packages-installation-and-loading">Packages installation and loading<a class="anchor" aria-label="anchor" href="#packages-installation-and-loading"></a>
</h3>
<p>Before installing <strong>envMRMEGAfm</strong> package, we load the
packages necessary for analysis, and check the version of each
package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#cran</span></span>
<span><span class="va">cran.pkgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'data.table'</span>, <span class="st">'doParallel'</span>,<span class="st">'dplyr'</span>,<span class="st">'foreach'</span>,<span class="st">'parallel'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load packages </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">cran.pkgs</span>, <span class="va">require</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="simulation-example">Simulation example<a class="anchor" aria-label="anchor" href="#simulation-example"></a>
</h2>
<p>In this simulated data example, we considered 5 continental African
populations which were collected from Phase 3 of the 1000 Genomes
Project <span class="citation">(<a href="#ref-10002015global">Consortium
et al. 2015</a>)</span>. The five populationss are coded as <span style="color: #31a354;">ESN</span>, <span style="color: #31a354;">GWD</span>, <span style="color: #31a354;">LWK</span>, <span style="color: #31a354;">MSL</span>, <span style="color: #31a354;">YRI</span>. Using the genotype data from each of
the five cohorts, we used <a href="https://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html" class="external-link">HAPGEN2</a>
<span class="citation">(<a href="#ref-su2011hapgen2">Su, Marchini, and
Donnelly 2011</a>)</span> software to generate individual-level genotype
data of each cohort, with a sample size of 10,000 for each cohort.
Within each population, we retained variants with MAF&gt;0.01,
regardless of whether the variant satisfies the MAF threshold in the
other African populations. Under this setting, the number of retained
variants differs across populations: 301 in <span style="color: #31a354;">ESN</span>, 291 in <span style="color: #31a354;">GWD</span>, 289 in <span style="color: #31a354;">LWK</span>, 299 in <span style="color: #31a354;">MSL</span> and 292 in <span style="color: #31a354;">YRI</span>. Each populations is then stratified
into female and male cohorts. When simulating 10 sex-stratified GWAS
files for a region in chromosome19: 45711510-45770083, two randomly
selected genetic variants, <span style="color: #964B00;">rs73939822</span> and <span style="color: #964B00;">rs60269219</span>, were designed as causal
variants (CVs) across all cohorts. Additionally, the corresponding
population-level LD correlation matrices are calculated and can be
loaded from the <strong>envMRMEGAfm</strong> package.</p>
<p>Fig 1 shows that the first two axes of genetic variation from
multi-dimensional scaling of the Euclidean distance matrix between 10
sex-stratified cohorts are sufficient to separate population groups from
different regions of Africa: east Africa (<span style="color: #31a354;">LWK</span>), west-central Africa (<span style="color: #31a354;">ESN</span>,<span style="color: #31a354;">YRI</span>) and west Africa (<span style="color: #31a354;">GWD</span>, <span style="color: #31a354;">MSL</span>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">'./figure/pcs_cnAFR.png'</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figure/pcs_cnAFR.png" alt="Figure 1. Axes of genetic variation separating five continental African cohorts" width="100%"><p class="caption">
Figure 1. Axes of genetic variation separating five continental African
cohorts
</p>
</div>
<p>Accounting for sex indicators and environmental exposures which
differ across all cohorts, we introduce sex and smoking status as the
two environmental covariates. Cohort-level environmental covariates are
constructed by taking the mean or proportion of each individual-level
covariate. Cohort-level sex and smoking covariates can be loaded using
the following codes:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">envMRMEGAfm</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sexsmk</span><span class="op">)</span></span>
<span><span class="va">sexsmk</span></span>
<span><span class="co">#&gt;    sex  smoke</span></span>
<span><span class="co">#&gt; 1    1 0.2120</span></span>
<span><span class="co">#&gt; 2    0 0.8314</span></span>
<span><span class="co">#&gt; 3    1 0.1890</span></span>
<span><span class="co">#&gt; 4    0 0.6750</span></span>
<span><span class="co">#&gt; 5    1 0.2332</span></span>
<span><span class="co">#&gt; 6    0 0.8014</span></span>
<span><span class="co">#&gt; 7    1 0.2652</span></span>
<span><span class="co">#&gt; 8    0 0.6554</span></span>
<span><span class="co">#&gt; 9    1 0.2608</span></span>
<span><span class="co">#&gt; 10   0 0.8810</span></span></code></pre></div>
<p>For simulating the (environment-adjusted) meta-analysis results, we
derived two axes of genetic variation (PCs) from a subset of variant
with MAF&gt;5% in across 10 sex-stratified African cohorts. The PCs can
be loaded from <strong>envMRMEGAfm</strong> package and are shown in the
following:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">envMRMEGAfm</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span></span>
<span><span class="va">pcs</span></span>
<span><span class="co">#&gt;               V1           V2</span></span>
<span><span class="co">#&gt; 1  -0.0035230264 -0.027745304</span></span>
<span><span class="co">#&gt; 2  -0.0038187733 -0.027451544</span></span>
<span><span class="co">#&gt; 3   0.0243106090  0.022216813</span></span>
<span><span class="co">#&gt; 4   0.0241666027  0.022483699</span></span>
<span><span class="co">#&gt; 5  -0.0429318096  0.018390796</span></span>
<span><span class="co">#&gt; 6  -0.0433946312  0.018995055</span></span>
<span><span class="co">#&gt; 7   0.0221012729  0.008319361</span></span>
<span><span class="co">#&gt; 8   0.0218724163  0.007775967</span></span>
<span><span class="co">#&gt; 9   0.0005143428 -0.021555609</span></span>
<span><span class="co">#&gt; 10  0.0007029969 -0.021429234</span></span></code></pre></div>
<p><span style="color: #3182bd;">Note</span>: - If PCs are not
available, <code>MR_mega_MDS</code> function implemented in <a href="https://SiruRooney.github.io/env.MRmega/" class="external-link">env.MRMEGA</a> package
is used for calculating PCs by inputting the file containing all GWAS
files location and the number of PCs.</p>
<div class="section level3">
<h3 id="data-pre-processing">Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h3>
<p>In <strong>envMRMEGAfm</strong> package, R function
<code>env_MR_MEGA_fm</code> is the main tool for fine-mapping genetic
associations for the specified traits. Before performing the
fine-mapping procedure, we need to make sure that all GWAS and LD
matrices are aligned to the same reference allele. Additionally, some
large-scaled LD matrices might contain redundant variants that are not
present in all GWAS files. Through data pre-processing, we simplify the
raw LD data by removing these redundant variants.</p>
<p>In practice, since both GWAS files and LD matrices contain a large
amount of genetic variants information, we need to create two input
files: one for all GWAS files locations and another for LD matrices
locations. The <code>read_gwasld</code> function is used for
pre-processing these raw data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.loc</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/esn_female.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/esn_male.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/gwd_female.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/gwd_male.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/lwk_female.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/lwk_male.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/msl_female.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/msl_male.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/yri_female.txt"</span>,</span>
<span>           <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/yri_male.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ld.loc</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/esn.txt"</span>,</span>
<span>         <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/gwd.txt"</span>,</span>
<span>         <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/lwk.txt"</span>,</span>
<span>         <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/msl.txt"</span>,</span>
<span>         <span class="st">"/Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/yri.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">LD</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"esn"</span>,<span class="st">"gwd"</span>,<span class="st">"lwk"</span>,<span class="st">"msl"</span>,<span class="st">"yri"</span><span class="op">)</span></span>
<span><span class="co">#Each African population is stratified into male and female cohorts.</span></span>
<span><span class="va">pop</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"esn.female"</span>,<span class="st">"esn.male"</span>,<span class="st">"gwd.female"</span>,<span class="st">"gwd.male"</span>,<span class="st">"lwk.female"</span>,<span class="st">"lwk.male"</span>,<span class="st">"msl.female"</span>,<span class="st">"msl.male"</span>,<span class="st">"yri.female"</span>,<span class="st">"yri.male"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gwasld</span><span class="op">=</span><span class="fu"><a href="../reference/read_gwasld.html">read_gwasld</a></span><span class="op">(</span><span class="va">gwas.loc</span>,<span class="va">ld.loc</span>,which.ld<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">LD</span>,each<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,cohort_name<span class="op">=</span><span class="va">pop</span>,ld_name<span class="op">=</span><span class="va">LD</span>,out_loc<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "i in filesize 1"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/esn_female.txt"</span></span>
<span><span class="co">#&gt; There are  301  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 2"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/esn_male.txt"</span></span>
<span><span class="co">#&gt; There are  301  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 3"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/gwd_female.txt"</span></span>
<span><span class="co">#&gt; There are  291  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 4"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/gwd_male.txt"</span></span>
<span><span class="co">#&gt; There are  291  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 5"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/lwk_female.txt"</span></span>
<span><span class="co">#&gt; There are  289  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 6"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/lwk_male.txt"</span></span>
<span><span class="co">#&gt; There are  289  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 7"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/msl_female.txt"</span></span>
<span><span class="co">#&gt; There are  299  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 8"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/msl_male.txt"</span></span>
<span><span class="co">#&gt; There are  299  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 9"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/yri_female.txt"</span></span>
<span><span class="co">#&gt; There are  292  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 10"</span></span>
<span><span class="co">#&gt; [1] "Reading gwas file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/gwasfile/yri_male.txt"</span></span>
<span><span class="co">#&gt; There are  294  SNPs in this GWAS file</span></span>
<span><span class="co">#&gt; [1] "i in filesize 1"</span></span>
<span><span class="co">#&gt; [1] "Reading LD file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/esn.txt"</span></span>
<span><span class="co">#&gt; [1] 301 301</span></span>
<span><span class="co">#&gt; [1] "i in filesize 2"</span></span>
<span><span class="co">#&gt; [1] "Reading LD file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/gwd.txt"</span></span>
<span><span class="co">#&gt; [1] 291 291</span></span>
<span><span class="co">#&gt; [1] "i in filesize 3"</span></span>
<span><span class="co">#&gt; [1] "Reading LD file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/lwk.txt"</span></span>
<span><span class="co">#&gt; [1] 289 289</span></span>
<span><span class="co">#&gt; [1] "i in filesize 4"</span></span>
<span><span class="co">#&gt; [1] "Reading LD file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/msl.txt"</span></span>
<span><span class="co">#&gt; [1] 299 299</span></span>
<span><span class="co">#&gt; [1] "i in filesize 5"</span></span>
<span><span class="co">#&gt; [1] "Reading LD file /Volumes/Mac/Stepwise_conditioning/Codes/env-MR-MEGAfm-Rpack/g1000_5afr_Rpack/LD/yri.txt"</span></span>
<span><span class="co">#&gt; [1] 293 293</span></span>
<span><span class="co">#&gt; [1] "Assign each GWAS file to the corresponding LD structure......"</span></span>
<span><span class="co">#&gt; [1] "matrix" "array" </span></span>
<span><span class="co">#&gt; [1] 301 301</span></span>
<span><span class="co">#&gt; [1] "matrix" "array" </span></span>
<span><span class="co">#&gt; [1] 291 291</span></span>
<span><span class="co">#&gt; [1] "matrix" "array" </span></span>
<span><span class="co">#&gt; [1] 289 289</span></span>
<span><span class="co">#&gt; [1] "matrix" "array" </span></span>
<span><span class="co">#&gt; [1] 299 299</span></span>
<span><span class="co">#&gt; [1] "matrix" "array" </span></span>
<span><span class="co">#&gt; [1] 293 293</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gwas.list</span><span class="op">=</span><span class="va">gwasld</span><span class="op">$</span><span class="va">gwas.list</span></span>
<span></span>
<span><span class="va">ld.list</span><span class="op">=</span><span class="va">gwasld</span><span class="op">$</span><span class="va">ld.list</span></span></code></pre></div>
<p>In the commands above, <code>read_gwasld</code> function outputs two
files: one containing the pre-processed GWAS files and another
containing the pre-processed LD matrices. These pre-processed GWAS files
and LD matrices can be loaded from the <strong>envMRMEGAfm</strong>
package.</p>
<p>More specifically, the above arguments present in
<code>read_gwasld</code> function are:</p>
<ul>
<li><p><code>gwas.loc</code>: A file contains a vector of length <span class="math inline">\(K_g\)</span>. Each component of the vector refers
to the location of GWAS file.</p></li>
<li><p><code>ld.loc</code>: A file contains a vector of length <span class="math inline">\(K_{ld}\)</span>. Each component of the vector
refers to the location of LD structure files.</p></li>
<li><p><code>which.ld</code>: A character vector of length <span class="math inline">\(K_g\)</span>. Each component of the vector
corresponds to one LD structure. The length of which.ld should equal to
the number of GWAS files. <span style="color: #3182bd;">Note</span>: the
argument <code>which.ld</code> is regarded as a link between GWAS files
and LD files. Each component of <code>which.ld</code> vector results
from one of the LD names.</p></li>
<li><p><code>cohort_name</code>: A character vector of length <span class="math inline">\(K_g\)</span>. Each component of the vector
corresponds to one GWAS file.</p></li>
<li><p><code>ld_name</code>: A character vector of length <span class="math inline">\(K_{ld}\)</span>. Each component of the vector
corresponds to one LD structure.</p></li>
<li><p><code>out_loc</code>: Path to save pre-processed GWAS files and
LD structure. By default, <code>out_loc</code>=NULL.</p></li>
<li><p><code>actual.geno</code>: An indicator to specify whether the
true cohort-level LD structure is provided. If we can obtain the real
cohort-level LD structures directly or derive LD structures from the
real individual-level genotype data of the involved cohorts, the
argument <code>actual.geno</code> is set to <code>TRUE</code>. If we
estimate LD approximation from the reference sample as a replacement,
<code>actual.geno</code> is set to <code>FALSE</code>. By default
<code>actual.geno</code>=FALSE.The default value is consistent for
GCTA-COJO setting.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="fine-mapping">Fine-mapping<a class="anchor" aria-label="anchor" href="#fine-mapping"></a>
</h3>
<p>In the simulation design where the allelic heterogeneity is
correlated with smoking status, we applied env-MR-MEGAfm method and
MR-MEGAfm to the simulated example respectively. Different from
env-MR-MEGAfm, we run MR-MEGAfm by setting the argument <code>env</code>
as <code>NULL</code>. In addition to some arguments appearing in
<code>read_gwasld</code> function, there are other arguments in
<code>env_MR_MEGA_fm</code> function. The additional arguments are
illustrated in the following:</p>
<ul>
<li><p><code>meta.file</code>: The meta-analysis results obtained from
the output of MR-MEGA method or env-MR-MEGA method. A dataframe contains
some mandatory columns where <strong>Preparation for
(env-)MR-MEGAfm</strong> section gives the further details.</p></li>
<li><p><code>ncores</code>: The number of cores which would be used for
running in parallel.</p></li>
<li><p><code>collinear</code>: A correlation threshold to filter out the
target SNP in high LD with the SNP set. If the squared multiple
correlation between the target SNP exceeds the threshold, such as 0.9,
the target SNP is ignored.</p></li>
<li><p><code>pvalue_cutoff</code>: A p-value threshold to identify
genetic associations. By default,
<code>pvalue_cutoff</code>=5e-8.</p></li>
<li><p><code>cred.thr</code>: Credible threshold for the credible set
for each selected potential SNP. By default, <code>cred.thr</code>=0.99
refers to 99% credible sets.</p></li>
</ul>
<p><span style="color: #3182bd;">Note</span>:</p>
<ul>
<li><p>In this simulated example, LD correlation matrix of each
population was derived from the true individual-level genotype data,
with the argument <code>actual.geno</code> set to <code>TRUE</code>.
However, in some cases, individual-level genotype data for each cohort
are poor, resulting in unavailable for calculation of LD matrices. In
such cases, the LD matrix can be estimated using a reference panel
(e.g. <a href="https://ctg.cncr.nl/software/MAGMA/ref_data/" class="external-link">1000
Genomes</a>) from a genetically similar population to the samples used
for genotype-phenotype association analysis. When the approximation of
LD matrix is applied, argument <code>actual.geno</code> should be set to
<code>FALSE</code>.</p></li>
<li><p>For (env-)MR-MEGAfm method, the key model selection strategy is
built upon (env-)MR-MEGA meta-analysis framework. As a result,
(env-)MR-MEGAfm method has the limitations regarding the number of PCs
(<span style="color: #3182bd;">T</span>) and the number of environment
exposures (<span style="color: #3182bd;">K</span>). Specifically, for
MR-MEGAfm, across (<span style="color: #3182bd;">K</span>) cohorts, the
number of axes of genetic variation (<span style="color: #3182bd;">T</span>) is constrained by <span style="color: #3182bd;">T&lt;K-2</span>. For env-MR-MEGAfm, the
limitation extends to the sum of the number of axes of genetic variation
(<span style="color: #3182bd;">T</span>) and the number of environmental
covariates (<span style="color: #3182bd;">S</span>), with the limitation
<span style="color: #3182bd;">T+S&lt;K-2</span>.</p></li>
</ul>
<p><strong>env-MR-MEGAfm</strong></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gwas.list</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "esn.female" "esn.male"   "gwd.female" "gwd.male"   "lwk.female"</span></span>
<span><span class="co">#&gt;  [6] "lwk.male"   "msl.female" "msl.male"   "yri.female" "yri.male"</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ld.list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "esn" "gwd" "lwk" "msl" "yri"</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">######################################env-MR-MEGA fine-mapping###################################</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="st">"Run env-MR-MEGA fine-mapping method"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RevolutionAnalytics/doparallel" class="external-link">doParallel</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: foreach</span></span>
<span><span class="co">#&gt; Loading required package: iterators</span></span>
<span><span class="co">#&gt; Loading required package: parallel</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">envmetafm.out</span><span class="op">=</span><span class="fu"><a href="../reference/env_MR_MEGA_fm.html">env_MR_MEGA_fm</a></span><span class="op">(</span><span class="va">gwas.list</span>,<span class="va">ld.list</span>,which.ld<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ld.list</span><span class="op">)</span>,each<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,meta.file<span class="op">=</span><span class="va">envMRMEGAout</span>,PCs<span class="op">=</span><span class="va">pcs</span>,env<span class="op">=</span><span class="va">sexsmk</span>,out_loc<span class="op">=</span><span class="cn">NULL</span>,ncores<span class="op">=</span><span class="fl">2</span>,collinear<span class="op">=</span><span class="fl">0.9</span>,pvalue_cutoff<span class="op">=</span><span class="fl">5e-8</span>,cred.thr<span class="op">=</span><span class="fl">0.99</span><span class="op">)</span></span>
<span><span class="co">#&gt; (Collinearity cutoff =  0.9 )</span></span>
<span><span class="co">#&gt; The number of cores are  2 , which are used for running in parallel</span></span>
<span><span class="co">#&gt; The threshold for credible set is set to  0.99 .</span></span>
<span><span class="co">#&gt; i.pop is  1 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  2 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  3 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  4 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  5 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  6 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  7 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  8 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  9 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  10 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; At the  1 iteration, the set of the selected SNPs include </span></span>
<span><span class="co">#&gt; [1] "rs73939822"</span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; At the  2 iteration, the set of the selected SNPs include </span></span>
<span><span class="co">#&gt; [1] "rs73939822" "rs60269219"</span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs73939822 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs60269219 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; There are  2  potential associated SNPs! </span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs73939822 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs60269219 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Run env-MR-MEGA fine-mapping method: 22.575 sec elapsed</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">envmetafm.out</span></span>
<span><span class="co">#&gt; $sel.set</span></span>
<span><span class="co">#&gt; [1] "rs73939822" "rs60269219"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tgCS.thr</span></span>
<span><span class="co">#&gt; $tgCS.thr$rs73939822</span></span>
<span><span class="co">#&gt;     MARKERNAME chisq_association ndf_association pvalue_association</span></span>
<span><span class="co">#&gt; 215 rs73939822          1972.823               5                  0</span></span>
<span><span class="co">#&gt;     chisq_anc_env_het ndf_chisq_anc_env_het pvalue_anc_env_het chisq_residual</span></span>
<span><span class="co">#&gt; 215          424.8579                     4       1.181767e-90       6.102112</span></span>
<span><span class="co">#&gt;     ndf_residual pvalue_residual   logBF chisq_env_het ndf_env_het</span></span>
<span><span class="co">#&gt; 215            5       0.2964094 980.655      413.6669           2</span></span>
<span><span class="co">#&gt;     pvalue_env_het chisq_anc_het ndf_anc_het pvalue_anc_het       PPs maxld</span></span>
<span><span class="co">#&gt; 215   1.490651e-90      2.552803           2      0.2790396 0.9987435     1</span></span>
<span><span class="co">#&gt;     cs.inc</span></span>
<span><span class="co">#&gt; 215      1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tgCS.thr$rs60269219</span></span>
<span><span class="co">#&gt;    MARKERNAME chisq_association ndf_association pvalue_association</span></span>
<span><span class="co">#&gt; 77 rs60269219          1701.090               5                  0</span></span>
<span><span class="co">#&gt; 76 rs61079153          1699.642               5                  0</span></span>
<span><span class="co">#&gt; 78 rs59356929          1691.853               5                  0</span></span>
<span><span class="co">#&gt;    chisq_anc_env_het ndf_chisq_anc_env_het pvalue_anc_env_het chisq_residual</span></span>
<span><span class="co">#&gt; 77          437.7434                     4       1.938165e-93       5.764427</span></span>
<span><span class="co">#&gt; 76          437.3153                     4       2.398521e-93       5.911945</span></span>
<span><span class="co">#&gt; 78          432.7377                     4       2.341041e-92       5.978292</span></span>
<span><span class="co">#&gt;    ndf_residual pvalue_residual    logBF chisq_env_het ndf_env_het</span></span>
<span><span class="co">#&gt; 77            5       0.3298202 844.7887      426.7314           2</span></span>
<span><span class="co">#&gt; 76            5       0.3148816 844.0645      426.5248           2</span></span>
<span><span class="co">#&gt; 78            5       0.3083369 840.1698      421.8789           2</span></span>
<span><span class="co">#&gt;    pvalue_env_het chisq_anc_het ndf_anc_het pvalue_anc_het         PPs maxld</span></span>
<span><span class="co">#&gt; 77   2.169937e-93      6.716531           2     0.03479557 0.666173442     1</span></span>
<span><span class="co">#&gt; 76   2.406172e-93      6.681750           2     0.03540597 0.322890947     1</span></span>
<span><span class="co">#&gt; 78   2.455613e-92      6.643331           2     0.03609267 0.006570714     1</span></span>
<span><span class="co">#&gt;    cs.inc</span></span>
<span><span class="co">#&gt; 77      1</span></span>
<span><span class="co">#&gt; 76      1</span></span>
<span><span class="co">#&gt; 78      1</span></span></code></pre></div>
<p><strong>MR-MEGAfm</strong></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">######################################MR-MEGA fine-mapping###################################</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="st">"Run env-MR-MEGA fine-mapping method"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RevolutionAnalytics/doparallel" class="external-link">doParallel</a></span><span class="op">)</span></span>
<span><span class="va">metafm.out</span><span class="op">=</span><span class="fu"><a href="../reference/env_MR_MEGA_fm.html">env_MR_MEGA_fm</a></span><span class="op">(</span><span class="va">gwas.list</span>,<span class="va">ld.list</span>,which.ld<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ld.list</span><span class="op">)</span>,each<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,meta.file<span class="op">=</span><span class="va">MRMEGAout</span>,PCs<span class="op">=</span><span class="va">pcs</span>,env<span class="op">=</span><span class="cn">NULL</span>,out_loc<span class="op">=</span><span class="cn">NULL</span>,ncores<span class="op">=</span><span class="fl">2</span>,collinear<span class="op">=</span><span class="fl">0.9</span>,pvalue_cutoff<span class="op">=</span><span class="fl">5e-8</span>,cred.thr<span class="op">=</span><span class="fl">0.99</span><span class="op">)</span></span>
<span><span class="co">#&gt; (Collinearity cutoff =  0.9 )</span></span>
<span><span class="co">#&gt; The number of cores are  2 , which are used for running in parallel</span></span>
<span><span class="co">#&gt; The threshold for credible set is set to  0.99 .</span></span>
<span><span class="co">#&gt; i.pop is  1 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  2 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  3 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  4 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  5 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  6 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  7 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  8 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  9 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; i.pop is  10 </span></span>
<span><span class="co">#&gt; Whether the inputted LD is cohort-matched LD FALSE </span></span>
<span><span class="co">#&gt; At the  1 iteration, the set of the selected SNPs include </span></span>
<span><span class="co">#&gt; [1] "rs73939822"</span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; At the  2 iteration, the set of the selected SNPs include </span></span>
<span><span class="co">#&gt; [1] "rs73939822" "rs60269219"</span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs73939822 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs60269219 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; There are  2  potential associated SNPs! </span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs73939822 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span>
<span><span class="co">#&gt; The target selected genetic variant is rs60269219 </span></span>
<span><span class="co">#&gt; [1] "All GWAS are assigned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "At the same, the correponding LD structures should be adjusted accordingly."</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Run env-MR-MEGA fine-mapping method: 20.796 sec elapsed</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">metafm.out</span></span>
<span><span class="co">#&gt; $sel.set</span></span>
<span><span class="co">#&gt; [1] "rs73939822" "rs60269219"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tgCS.thr</span></span>
<span><span class="co">#&gt; $tgCS.thr$rs73939822</span></span>
<span><span class="co">#&gt;      MARKERNAME chisq_association ndf_association pvalue_association</span></span>
<span><span class="co">#&gt; 215  rs73939822          1559.156               3                  0</span></span>
<span><span class="co">#&gt; 222 rs143024846          1551.294               3                  0</span></span>
<span><span class="co">#&gt;     chisq_anc_het ndf_chisq_anc_het pvalue_anc_het chisq_residual ndf_residual</span></span>
<span><span class="co">#&gt; 215      11.19101                 2    0.003714517       419.7690            7</span></span>
<span><span class="co">#&gt; 222      10.05222                 2    0.006564303       414.2482            7</span></span>
<span><span class="co">#&gt;     pvalue_residual    logBF        PPs maxld cs.inc</span></span>
<span><span class="co">#&gt; 215    1.370481e-86 776.1242 0.98075059     1      1</span></span>
<span><span class="co">#&gt; 222    2.096064e-85 772.1933 0.01924941     1      1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tgCS.thr$rs60269219</span></span>
<span><span class="co">#&gt;    MARKERNAME chisq_association ndf_association pvalue_association</span></span>
<span><span class="co">#&gt; 77 rs60269219          1274.359               3      5.387582e-276</span></span>
<span><span class="co">#&gt; 76 rs61079153          1273.117               3      1.001923e-275</span></span>
<span><span class="co">#&gt; 78 rs59356929          1269.974               3      4.818463e-275</span></span>
<span><span class="co">#&gt; 82 rs57294488          1268.921               3      8.154631e-275</span></span>
<span><span class="co">#&gt;    chisq_anc_het ndf_chisq_anc_het pvalue_anc_het chisq_residual ndf_residual</span></span>
<span><span class="co">#&gt; 77      11.01200                 2    0.004062321       432.4959            7</span></span>
<span><span class="co">#&gt; 76      10.79051                 2    0.004538063       432.4367            7</span></span>
<span><span class="co">#&gt; 78      10.85875                 2    0.004385831       427.8572            7</span></span>
<span><span class="co">#&gt; 82      10.98314                 2    0.004121372       428.0893            7</span></span>
<span><span class="co">#&gt;    pvalue_residual    logBF        PPs maxld cs.inc</span></span>
<span><span class="co">#&gt; 77    2.544181e-89 633.7256 0.58308761     1      1</span></span>
<span><span class="co">#&gt; 76    2.619670e-89 633.1047 0.31338763     1      1</span></span>
<span><span class="co">#&gt; 78    2.518646e-88 631.5329 0.06508363     1      1</span></span>
<span><span class="co">#&gt; 82    2.245652e-88 631.0064 0.03844113     1      1</span></span></code></pre></div>
<p>The results from both MR-MEGAfm and env-MR-MEGAfm comprise two main
parts:</p>
<ul>
<li><p><code>sel.set</code>: The selected potential associated SNP
set.</p></li>
<li><p><code>tgCS.thr</code>: A list of the credible sets of the
selected SNPs under the specified threshold. By default, the specified
threshold is set to 0.99. Each component of the list is a dataframe
containing the following columns:</p></li>
<li><p><code>MARKERNAME</code>: unique marker identification across all
cohorts.</p></li>
<li><p><code>chisq_association</code>, <code>ndf_association</code> and
<code>pvalue_association</code>: test of the null hypothesis of no
association of each variant. <code>chisq_</code>, <code>ndf_</code> and
<code>pvalue_</code> refer to chisq value, number of degrees of freedom
and p-values of the corresponding test.</p></li>
<li><p><code>chisq_residual</code>, <code>ndf_residual</code> and
<code>pvalue_residual</code>: test of residual heterogeneity.</p></li>
<li><p><code>logBF</code>: log of Bayesian Factor.</p></li>
<li><p><code>PPs</code>: Posterior probability.</p></li>
<li><p><code>max_ld</code>: The maximum of LD value between each target
selected variant and the variants in the credible set.</p></li>
<li><p><code>cs.inc</code>: The indicator for determining whether the
variants should be included in the credible sets based on the specified
credible threshold.</p></li>
</ul>
<p>For MR-MEGAfm,</p>
<ul>
<li>
<code>chisq_anc_het</code>, <code>ndf_anc_het</code> and
<code>pvalue_anc_het</code>: test of ancestral heterogeneity.</li>
</ul>
<p>For env-MR-MEGAfm, results contain tests which are different from
MR-MEGAfm.</p>
<ul>
<li><p><code>chisq_anc_env_het</code>, <code>ndf_anc_env_het</code> and
<code>pvalue_anc_env_het</code>: test of ancestral and environmental
heterogeneity.</p></li>
<li><p><code>chisq_anc_het</code>, <code>ndf_anc_het</code> and
<code>pvalue_anc_het</code>: test of allelic heterogeneity due to
ancestry alone.</p></li>
<li><p><code>chisq_env_het</code>, <code>ndf_env_het</code> and
<code>pvalue_env_het</code>: test of allelic heterogeneity due to
environment alone.</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-10002015global" class="csl-entry">
Consortium, 1000 Genomes Project et al. 2015. <span>“A Global Reference
for Human Genetic Variation.”</span> <em>Nature</em> 526 (7571): 68.
</div>
<div id="ref-magi2017trans" class="csl-entry">
Mägi, Reedik, Momoko Horikoshi, Tamar Sofer, Anubha Mahajan, Hidetoshi
Kitajima, Nora Franceschini, Mark I McCarthy, T2D-GENES Consortium
COGENT-Kidney Consortium, and Andrew P Morris. 2017. <span>“Trans-Ethnic
Meta-Regression of Genome-Wide Association Studies Accounting for
Ancestry Increases Power for Discovery and Improves Fine-Mapping
Resolution.”</span> <em>Human Molecular Genetics</em> 26 (18): 3639–50.
</div>
<div id="ref-su2011hapgen2" class="csl-entry">
Su, Zhan, Jonathan Marchini, and Peter Donnelly. 2011. <span>“HAPGEN2:
Simulation of Multiple Disease SNPs.”</span> <em>Bioinformatics</em> 27
(16): 2304–5.
</div>
<div id="ref-wang2024accounting" class="csl-entry">
Wang, Siru, Oyesola O Ojewunmi, Abram Kamiza, Michele Ramsay, Andrew P
Morris, Tinashe Chikowore, Segun Fatumo, and Jennifer L Asimit. 2024.
<span>“Accounting for Heterogeneity Due to Environmental Sources in
Meta-Analysis of Genome-Wide Association Studies.”</span>
<em>Communications Biology</em> 7 (1): 1512.
</div>
<div id="ref-yang2012conditional" class="csl-entry">
Yang, Jian, Teresa Ferreira, Andrew P Morris, Sarah E Medland, Genetic
Investigation of ANthropometric Traits (GIANT) Consortium, DIAbetes
Genetics Replication And Meta-analysis (DIAGRAM) Consortium, Pamela AF
Madden, et al. 2012. <span>“Conditional and Joint Multiple-SNP Analysis
of GWAS Summary Statistics Identifies Additional Variants Influencing
Complex Traits.”</span> <em>Nature Genetics</em> 44 (4): 369–75.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siru Wang, Jennifer Asimit, Andrew P Morris.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
