<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>(env-)MR-MEGA fine-mapping approach (env-MR-MEGAfm) — env_MR_MEGA_fm • envMRMEGAfm</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="(env-)MR-MEGA fine-mapping approach (env-MR-MEGAfm) — env_MR_MEGA_fm"><meta property="og:description" content="Fine-mapping genetic associations across multiple cohorts, allowing for multiple causal variants."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">envMRMEGAfm</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/envMRMEGAfm.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>(env-)MR-MEGA fine-mapping approach (env-MR-MEGAfm)</h1>
    
    <div class="hidden name"><code>env_MR_MEGA_fm.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fine-mapping genetic associations across multiple cohorts, allowing for multiple causal variants.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">env_MR_MEGA_fm</span><span class="op">(</span></span>
<span>  <span class="va">gwas.list</span>,</span>
<span>  <span class="va">ld.list</span>,</span>
<span>  <span class="va">which.ld</span>,</span>
<span>  <span class="va">meta.file</span>,</span>
<span>  <span class="va">PCs</span>,</span>
<span>  <span class="va">env</span>,</span>
<span>  out_loc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  collinear <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  pvalue_cutoff <span class="op">=</span> <span class="fl">5e-08</span>,</span>
<span>  cred.thr <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>  actual.geno <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gwas.list</dt>
<dd><p>A list of length K_g which contains the pre-processed GWAS files. Each component contains one GWAS file comprising these required columns:
"MARKERNAME","CHROMOSOME","POSITION","EA","NEA","EAF","N","BETA","SE".</p></dd>


<dt>ld.list</dt>
<dd><p>A list of length K_ld which contains the pre-processed LD structures. Each component contains one LD structure.</p></dd>


<dt>which.ld</dt>
<dd><p>A character vector of length K_g. Each component of the vector corresponds to one LD structure. The length of which.ld should equal to the number of gwas files.</p></dd>


<dt>meta.file</dt>
<dd><p>A data frame obtained from the output of MR-MEGA method or env-MR-MEGA method.</p></dd>


<dt>PCs</dt>
<dd><p>The axes of genetic variation, which can also be called the principal components (PCs). Each row refers to one population and each column refers to one axes of genetic variation.
The order of matrix row in <code>PCs</code> should be consistent with the order of gwas names.</p></dd>


<dt>env</dt>
<dd><p>The matrix which contains the study-level environmental factors. Each row refers to on population and each column refers to one environment covariate.
The order of matrix row in <code>env</code> should be consistent with the order of gwas names. If MR-MEGAfm is utilized, env=NULL.</p></dd>


<dt>out_loc</dt>
<dd><p>Path to save the output of (env-)MR-MEGAfm. By default, out_loc=NULL.</p></dd>


<dt>ncores</dt>
<dd><p>The the number of cores which would be used for running in parallel.</p></dd>


<dt>collinear</dt>
<dd><p>A threshold to filter out the target SNP in high LD with the SNP set. If the squared multiple correlation between the target SNP exceeds the threshold,
such as 0.9, the target SNP is ignored.</p></dd>


<dt>pvalue_cutoff</dt>
<dd><p>A cutoff p-value to identify the potential associated SNP with the smallest p-value exceeding the cutoff p-value. By default, pvalue_cutoff=5e-8.</p></dd>


<dt>cred.thr</dt>
<dd><p>Credible threshold for the credible set for each selected potential SNP. By default, cred.thr=0.99 refers to 99% credible sets.</p></dd>


<dt>actual.geno</dt>
<dd><p>An indicator to specify whether the true cohort-level LD structure is applied. If actual.geno is TRUE, the inputted LD structures are derived from the true individual-level genotype data. By default, actual.geno=FALSE is recommended in GCTA-COJO software.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of length 2. One component contains the selected potential associated SNP. Another component contains their credible sets with the specified thresholds. By default, the credible threshold is set to 0.99.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Siru Wang</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siru Wang, Jennifer Asimit, Andrew P Morris.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

