<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>(env-)MR-MEGA fine-mapping approach (env-MR-MEGAfm) — env_MR_MEGA_fm • envMRMEGAfm</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="(env-)MR-MEGA fine-mapping approach (env-MR-MEGAfm) — env_MR_MEGA_fm"><meta name="description" content="Fine-mapping genetic associations across multiple cohorts, allowing for multiple causal variants."><meta property="og:description" content="Fine-mapping genetic associations across multiple cohorts, allowing for multiple causal variants."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">envMRMEGAfm</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/envMRMEGAfm.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>(env-)MR-MEGA fine-mapping approach (env-MR-MEGAfm)</h1>

      <div class="d-none name"><code>env_MR_MEGA_fm.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Fine-mapping genetic associations across multiple cohorts, allowing for multiple causal variants.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">env_MR_MEGA_fm</span><span class="op">(</span></span>
<span>  <span class="va">gwas.list</span>,</span>
<span>  <span class="va">ld.list</span>,</span>
<span>  <span class="va">which.ld</span>,</span>
<span>  <span class="va">meta.file</span>,</span>
<span>  <span class="va">PCs</span>,</span>
<span>  <span class="va">env</span>,</span>
<span>  out_loc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  collinear <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  pvalue_cutoff <span class="op">=</span> <span class="fl">5e-08</span>,</span>
<span>  cred.thr <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>  actual.geno <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gwas-list">gwas.list<a class="anchor" aria-label="anchor" href="#arg-gwas-list"></a></dt>
<dd><p>A list of length K_g which contains the pre-processed GWAS files. Each component contains one GWAS file comprising these required columns:
"MARKERNAME","CHROMOSOME","POSITION","EA","NEA","EAF","N","BETA","SE".</p></dd>


<dt id="arg-ld-list">ld.list<a class="anchor" aria-label="anchor" href="#arg-ld-list"></a></dt>
<dd><p>A list of length K_ld which contains the pre-processed LD structures. Each component contains one LD structure.</p></dd>


<dt id="arg-which-ld">which.ld<a class="anchor" aria-label="anchor" href="#arg-which-ld"></a></dt>
<dd><p>A character vector of length K_g. Each component of the vector corresponds to one LD structure. The length of which.ld should equal to the number of gwas files.</p></dd>


<dt id="arg-meta-file">meta.file<a class="anchor" aria-label="anchor" href="#arg-meta-file"></a></dt>
<dd><p>A data frame obtained from the output of MR-MEGA method or env-MR-MEGA method.</p></dd>


<dt id="arg-pcs">PCs<a class="anchor" aria-label="anchor" href="#arg-pcs"></a></dt>
<dd><p>The axes of genetic variation, which can also be called the principal components (PCs). Each row refers to one population and each column refers to one axes of genetic variation.
The order of matrix row in <code>PCs</code> should be consistent with the order of gwas names.</p></dd>


<dt id="arg-env">env<a class="anchor" aria-label="anchor" href="#arg-env"></a></dt>
<dd><p>The matrix which contains the study-level environmental factors. Each row refers to on population and each column refers to one environment covariate.
The order of matrix row in <code>env</code> should be consistent with the order of gwas names. If MR-MEGAfm is utilized, env=NULL.</p></dd>


<dt id="arg-out-loc">out_loc<a class="anchor" aria-label="anchor" href="#arg-out-loc"></a></dt>
<dd><p>Path to save the output of (env-)MR-MEGAfm. By default, out_loc=NULL.</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>The the number of cores which would be used for running in parallel.</p></dd>


<dt id="arg-collinear">collinear<a class="anchor" aria-label="anchor" href="#arg-collinear"></a></dt>
<dd><p>A threshold to filter out the target SNP in high LD with the SNP set. If the squared multiple correlation between the target SNP exceeds the threshold,
such as 0.9, the target SNP is ignored.</p></dd>


<dt id="arg-pvalue-cutoff">pvalue_cutoff<a class="anchor" aria-label="anchor" href="#arg-pvalue-cutoff"></a></dt>
<dd><p>A cutoff p-value to identify the potential associated SNP with the smallest p-value exceeding the cutoff p-value. By default, pvalue_cutoff=5e-8.</p></dd>


<dt id="arg-cred-thr">cred.thr<a class="anchor" aria-label="anchor" href="#arg-cred-thr"></a></dt>
<dd><p>Credible threshold for the credible set for each selected potential SNP. By default, cred.thr=0.99 refers to 99% credible sets.</p></dd>


<dt id="arg-actual-geno">actual.geno<a class="anchor" aria-label="anchor" href="#arg-actual-geno"></a></dt>
<dd><p>An indicator to specify whether the true cohort-level LD structure is applied. If actual.geno is TRUE, the inputted LD structures are derived from the true individual-level genotype data. By default, actual.geno=FALSE is recommended in GCTA-COJO software.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of length 2. One component contains the selected potential associated SNP. Another component contains their credible sets with the specified thresholds. By default, the credible threshold is set to 0.99.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Siru Wang</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Siru Wang, Jennifer Asimit, Andrew P Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

