<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Conditional analysis of the linear regression model — meta_sel_condest • envMRMEGAfm</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Conditional analysis of the linear regression model — meta_sel_condest"><meta name="description" content="Estimate the allelic effect size and the standard errors of the target SNP conditional on the SNP set."><meta property="og:description" content="Estimate the allelic effect size and the standard errors of the target SNP conditional on the SNP set."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">envMRMEGAfm</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/envMRMEGAfm.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Conditional analysis of the linear regression model</h1>

      <div class="d-none name"><code>meta_sel_condest.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Estimate the allelic effect size and the standard errors of the target SNP conditional on the SNP set.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">meta_sel_condest</span><span class="op">(</span></span>
<span>  <span class="va">add.gwas</span>,</span>
<span>  <span class="va">sel.gwas</span>,</span>
<span>  <span class="va">addsel.ld</span>,</span>
<span>  <span class="va">sel.ld</span>,</span>
<span>  <span class="va">Vp.med</span>,</span>
<span>  <span class="va">collinear</span>,</span>
<span>  <span class="va">actual.geno</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-add-gwas">add.gwas<a class="anchor" aria-label="anchor" href="#arg-add-gwas"></a></dt>
<dd><p>A dataframe which contains the GWAS information for the target SNP. GWAS information comprise "MARKERNAME","CHROMOSOME","POSITION","EA","NEA","EAF","N","BETA","SE".</p></dd>


<dt id="arg-sel-gwas">sel.gwas<a class="anchor" aria-label="anchor" href="#arg-sel-gwas"></a></dt>
<dd><p>A dataframe which contains the GWAS information for SNP set.</p></dd>


<dt id="arg-addsel-ld">addsel.ld<a class="anchor" aria-label="anchor" href="#arg-addsel-ld"></a></dt>
<dd><p>A LD matrix containing the correlation matrix between the target SNP and the SNP set.</p></dd>


<dt id="arg-sel-ld">sel.ld<a class="anchor" aria-label="anchor" href="#arg-sel-ld"></a></dt>
<dd><p>A LD matrix containing the correlation between the SNP set.</p></dd>


<dt id="arg-vp-med">Vp.med<a class="anchor" aria-label="anchor" href="#arg-vp-med"></a></dt>
<dd><p>The median of variance of phenotype. Refer to GCTA-COJO.</p></dd>


<dt id="arg-collinear">collinear<a class="anchor" aria-label="anchor" href="#arg-collinear"></a></dt>
<dd><p>A threshold to filter out the target SNP in high LD with the SNP set. If the squared multiple correlation between the target SNP exceeds the threshold,
such as 0.9, the target SNP is ignored.</p></dd>


<dt id="arg-actual-geno">actual.geno<a class="anchor" aria-label="anchor" href="#arg-actual-geno"></a></dt>
<dd><p>An indicator to specify whether the true cohort-level LD structure is applied. If actual.geno is TRUE, the inputted LD structures are derived from the true individual-level genotype data.
If actual.geno is FALSE, the the inputted LD structures are derived from the approximation LD strucure.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Output a data.frame containing the estimated allelic effect and  the inverse of squared standard error for the target SNP.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Siru Wang</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Siru Wang, Jennifer Asimit, Andrew P Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

